app.py

```diff
@@ -32,21 +32,79 @@
     'upstream/Rover-4.2' : 'Rover 4.2 stable'
 }
 default_branch = 'upstream/master'
-CURR_BRANCH = default_branch
 
-def get_boards_from_ardupilot_tree(branch):
+def get_vehicles():
+    return VEHICLES
+
+def get_default_vehicle():
+    return default_vehicle
+
+def get_branches():
+    return BRANCHES
+
+def get_default_branch():
+    return default_branch
+
+# LOCKS
+queue_lock = Lock()
+head_lock = Lock()  # lock git HEAD, i.e., no branch change until this lock is released
+
+def is_valid_branch(branch):
+    return get_branches().get(branch) is not None
+
+def get_git_hash(branch):
+    return subprocess.check_output(['git', 'rev-parse', branch], cwd=sourcedir, encoding='utf-8').rstrip()
+
+def on_branch(branch):
+    git_hash_target = get_git_hash(branch)
+    app.logger.info("Expected branch git-hash '%s'" % git_hash_target)
+    git_hash_current = get_git_hash('HEAD')
+    app.logger.info("Current branch git-hash '%s'" % git_hash_current)
+    return git_hash_target == git_hash_current
+
+def checkout_branch(targetBranch, s_dir, fetch_and_reset=False, temp_branch_name=None):
+    '''checkout to given branch and return the git hash'''
+    # Note: remember to acquire head_lock before calling this method
+    if not is_valid_branch(targetBranch):
+        app.logger.error("Checkout requested for an invalid branch")
+        return None 
+    remote =  targetBranch.split('/', 1)[0]
+    branch_name = targetBranch.split('/', 1)[1]
+    if not on_branch(targetBranch):
+        app.logger.info("Checking out to %s branch" % targetBranch)
```

Comment:
Is there some reason not to check the branches out in a detached-head state rather than creating a local branch?

i.e. `git checkout <remotename>/<branchname>` rather than just `git  checkout branchname`

Reply:
We can't clone a remote branch to a directory using `git clone --single-branch` directly. It has to be a local branch to clone it. So we need to create a temp branch and then clone it.

Reply:
If we only want to checkout (without cloning) that happens in detached head state. The `tagetBranch` is already in the format `<remotename>/<branchname>`

Reply:
It was failing sometimes for few branches due to same branch names in multiple remotes.

Reply:
When I did checkout using `git checkout Copter-4.2`, It threw the error saying 
```
hint: 'Copter-4.2' matched more than one remote tracking branch.
hint: We found 2 remotes with a reference that matched. So we fell back
hint: on trying to resolve the argument as a path, but failed there too!
hint: 
hint: If you meant to check out a remote tracking branch on, e.g. 'origin',
hint: you can do so by fully qualifying the name with the --track option:
hint: 
hint:     git checkout --track origin/<name>
hint: 
hint: If you'd like to always have checkouts of an ambiguous <name> prefer
hint: one remote, e.g. the 'origin' remote, consider setting
hint: checkout.defaultRemote=origin in your config.
```

Reply:
Then shouldn't using the detached head method work well given you're giving it the remote as well?

Reply:
Yes, we are checking out in detached head state only. ```subprocess.run(['git', 'checkout', targetBranch], cwd=s_dir, shell=False)```
Here, target branch has remote and branch name both.

The need for copying remote branch to a temporary local branch only comes when we need to clone that branch to a directory. The remote branch can't be copied to a directory directly. Hence, we first copy that branch to a new temp branch using `git switch -c` and then proceed with cloning.

Hence, if we are not looking for cloning and only want to checkout to a remote branch, we only do `git checkout targetBranch` which happens in detached head state only.